File: optionals/med-4.0.0/tests/f/test3.f
C*  This file is part of MED.
C*
C*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
C*  MED is free software: you can redistribute it and/or modify
C*  it under the terms of the GNU Lesser General Public License as published by
C*  the Free Software Foundation, either version 3 of the License, or
C*  (at your option) any later version.
C*
C*  MED is distributed in the hope that it will be useful,
C*  but WITHOUT ANY WARRANTY; without even the implied warranty of
C*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
C*  GNU Lesser General Public License for more details.
C*
C*  You should have received a copy of the GNU Lesser General Public License
C*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
C*

C ******************************************************************************
C * - Nom du fichier : test3.f
C *
C * - Description : lecture des informations sur les maillages dans un fichier
C*                  MED.
C *
C ******************************************************************************
      program test3
C     
      implicit none
      include 'med.hf'
C
C
      integer*8     fid
      integer       cret,cres,type,cnu
      character*64  maa
      character*80  nomu
      character*200 desc
      integer       nmaa,i,mdim,edim,nstep,stype,atype
C  ** chgt de dim 2->3 car le fichier dump.ref/test2.med en 2.3.6 est utilisé comme référence
C  ** (il contient un maillage de dimension 3 et un espace induit de dimension 3
C  **   car pas de coordonée stockée) 
C  ** dans 2.3v3.0 qui utilise ce test3 en v3.0 qui défini nomcoo et unicoo en dimension 2
C      character*16 nomcoo(2)
C      character*16 unicoo(2)
      character*16 nomcoo(3)
      character*16 unicoo(3)
      character*16 dtunit
      integer      maa1exist,maa4exist
      
C ** Ouverture du fichier en lecture seule
      call mfiope(fid,'test2.med',MED_ACC_RDONLY, cret)
      print *,cret
      if (cret .ne. 0 ) then
         print *,'Erreur ouverture du fichier en lecture'
         call efexit(-1)
      endif      

C ** Test de la présence d'un maillage
      call mfioex(fid,MED_MESH,"maa1", maa1exist, cret)
      print *,cret
      if (cret .ne. 0 ) then
         print *,'Erreur de test de présence de maillage'
         call efexit(-1)
      endif      
      print *,"Maillage maa1 existe : ",maa1exist

      call mfioex(fid,MED_MESH,"maa4", maa4exist, cret)
      print *,cret
      if (cret .ne. 0 ) then
         print *,'Erreur de test de présence de maillage'
         call efexit(-1)
      endif      
      print *,"Maillage maa4 existe : ",maa4exist

C ** lecture du nombre de maillage                      **
      call mmhnmh(fid,nmaa,cret)
      print *,cret
      if (cret .ne. 0 ) then
         print *,'Erreur lecture du nombre de maillage'
         call efexit(-1)
      endif      
      print *,'Nombre de maillages = ',nmaa

C ** lecture des infos sur les maillages : **
C ** - nom, dimension, type,description
C ** - options : nom universel, dimension de l'espace
      do i=1,nmaa  
         call mmhmii(fid,i,maa,edim,mdim,type,desc,
     &               dtunit,stype,nstep,atype,
     &               nomcoo,unicoo,cret)
         call mmhunr(fid,maa,nomu,cnu)
         print *,cret
         if (cret .ne. 0 ) then
            print *,'Erreur acces au maillage'
            call efexit(-1)
         endif      
         print '(A,I1,A,A4,A,I1,A,A65,A65)','maillage '
     &        ,i,' de nom ',maa,' et de dimension ',mdim,
     &        ' de description ',desc
         if (type.eq.MED_UNSTRUCTURED_MESH) then
            print *,'Maillage non structure'
         else
            print *,'Maillage structure'
         endif
         print *,'Dimension espace   ', edim
         print *,'Dimension maillage ', mdim
         if (cnu.eq.0) then
            print *,'Nom universel : ',nomu
         else
            print *,'Pas de nom universel'
         endif
         print *,'dt unit = ', dtunit
         print *,'sorting type =', stype
         print *,'number of computing step =', nstep
         print *,'coordinates axis type =', atype
         print *,'coordinates axis name =', nomcoo(1),nomcoo(2)
         print *,'coordinates axis units =', unicoo(1),unicoo(2)
      enddo         
         
C **  fermeture du fichier
      call mficlo(fid,cret)
      print *,cret
      if (cret .ne. 0 ) then
         print *,'Erreur fermeture du fichier'
         call efexit(-1)
      endif      
C
      end 

File: optionals/med-4.0.0/tests/c/test3.c
/*  This file is part of MED.
 *
 *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
 *  MED is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MED is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
 */


/******************************************************************************
 * - Nom du fichier : test3.c
 *
 * - Description : lecture des informations sur les maillages d'un fichier MED.
 *
 *****************************************************************************/

#include <med.h>
#define MESGERR 1
#include "med_utils.h"
#include <string.h>

#ifdef DEF_LECT_ECR
#define MODE_ACCES MED_ACC_RDWR
#elif DEF_LECT_AJOUT
#define MODE_ACCES MED_ACC_RDEXT
#else
#define MODE_ACCES MED_ACC_CREAT
#endif


#define OBJEXIST(oname)\
 if ( MEDfileObjectExist(fid,MED_MESH, #oname ,& oname##exist) < 0) {	\
    MESSAGE("Erreur de test de présence du maillage "#oname);	\
    return -1;							\
  }								\
  if ( oname##exist) { MESSAGE("Le maillage "#oname" existe.");  } else \
                     { MESSAGE("Le maillage "#oname" n'existe pas.");}


int main (int argc, char **argv)


{
  med_err ret = 0;
  med_idt fid = 0;
  med_int nmaa=0,mdim=0,sdim=0,nstep=0;
  int i=0;
  /*Pour les outils de type memchecker */
/*   char maa[MED_NAME_SIZE+1]=""; */
/*   char nomu[MED_LNAME_SIZE+1]=""; */
/*   char des [MED_COMMENT_SIZE+1]=""; */
/*   char dtunit[MED_SNAME_SIZE+1]=""; */
  char *maa   ;
  char *nomu  ;
  char *des   ;
  char *dtunit;
  char *axisname=0,*axisunit=0;
  med_mesh_type     meshtype;
  med_err           inomu;
  med_sorting_type  sortingtype;
  med_axis_type     axistype;
  med_bool          maa1exist=MED_FALSE,maa2exist=MED_FALSE;
  med_bool          maa3exist=MED_FALSE,maa4exist=MED_FALSE;
  
  /*Pour les outils de type memchecker */
  maa    = (char *) malloc(sizeof(char)*(MED_NAME_SIZE+1   ));
  nomu   = (char *) malloc(sizeof(char)*(MED_LNAME_SIZE+1  ));
  des    = (char *) malloc(sizeof(char)*(MED_COMMENT_SIZE+1));
  dtunit = (char *) malloc(sizeof(char)*(MED_SNAME_SIZE+1  ));

  /* Ouverture du fichier "test2.med" en lecture seule */
  fid = MEDfileOpen("test2.med",MED_ACC_RDONLY);
  if (fid < 0) {
    MESSAGE("Erreur a l'ouverture du fichier test2.med");
    return -1;
  }

  /* Teste la présence des maillages */
  OBJEXIST(maa1);  OBJEXIST(maa2);  OBJEXIST(maa3);  OBJEXIST(maa4);
 
  /* Lecture du nombre de maillage dans le fichier */
  nmaa = MEDnMesh(fid);
  if (nmaa < 0) {
    MESSAGE("Erreur a la lecture du nombre de maillage");
    return  -1;
  }


  /* Boucle sur tous les maillages, pour chaque maillage, on lit :
     - Le nom.
     - Le type
     - La dimension
     - La description
     - La dimension de l'espace si elle existe
     - Le nom universel s'il existe
  */
  printf("- Nombre de maillage dans test2.med = "IFORMAT"\n",nmaa);

  for (i=0;i< nmaa;i++) {
    /* lecture des informations */

    if ((sdim=MEDmeshnAxis(fid, i+1)) <0) {
      MESSAGE("Erreur a la lecture de la dimension de l'espace du maillage :");
      SSCRUTE(maa);
      ret = -1;
    }
    axisname  = (char*) malloc(MED_SNAME_SIZE*sdim+1);
    axisunit  = (char*) malloc(MED_SNAME_SIZE*sdim+1);

    if (MEDmeshInfo(fid,i+1, maa, &sdim, &mdim,  &meshtype, des, dtunit, &sortingtype, &nstep,
		    &axistype, axisname, axisunit) < 0) {
      MESSAGE("Erreur a la lecture des informations du maillage :"); SSCRUTE(maa);
      ret = -1;
    }
    /* lecture du nom universel */
    inomu = MEDmeshUniversalNameRd(fid,maa,nomu);
    /* affichage des donnees lues */
    if (inomu < 0)
      printf("maillage %d de nom %s, de dimension "IFORMAT" \n",i+1,maa,mdim);
    else
      printf("maillage %d de nom %s, de dimension "IFORMAT" et de nom univ. %s\n",i+1,maa,mdim,nomu);

    printf("La dimension de l'espace est "IFORMAT" \n",sdim);
    if (meshtype == MED_STRUCTURED_MESH)
      printf("Il s'agit d'un maillage structure \n");
    else
      printf("Il s'agit d'un maillage non structure \n");
    printf("Description associee au maillage : %s \n\n",des);
    printf("\t -Noms des axes : %s\n",  axisname);
    printf("\t -Unités des axes : %s\n",axisunit);
    printf("\t -Type de repère : %d\n", axistype);
    printf("\t -Nombre d'étape de calcul : "IFORMAT"\n",nstep);
    printf("\t -Unité des dates : %s\n",dtunit);

    free(axisname);
    free(axisunit);
  }

  free(  maa    );
  free(  nomu   );
  free(  des    );
  free(  dtunit );


  /* Fermeture du fichier */
  if ( MEDfileClose(fid) < 0) {
    MESSAGE("Erreur a la fermeture du fichier test2.med");
    return -1;
  }

  return ret;
}




